cli/file_watcher.rs

```diff
@@ -90,11 +112,95 @@ where
         );
       },
       _ = func => {},
+    };
+
+    if !is_file_changed {
+      info!(
+        "{} {} finished! Restarting on file change...",
+        colors::intense_blue("Watcher"),
+        job_name,
+      );
+      debounce.next().await;
+      info!(
+        "{} File change detected! Restarting!",
+        colors::intense_blue("Watcher"),
+      );
+    }
+  }
+}
+
+/// This function adds watcher functionality to subcommands like `run` or `bundle`.
+/// The difference from [`watch_func`] is that this does depend on [`ModuleGraph`].
+///
+/// - `module_resolver` is used for both resolving file paths to be watched at every restarting
+/// of the watcher and building [`ModuleGraph`] or [`ModuleSpecifier`] which will then be passed
+/// to `operation`.
+///
+/// - `operation` is the actual operation we want to run every time the watcher detects file
+/// changes. For example, in the case where we would like to bundle, then `operation` would
+/// have the logic for it like doing bundle with the help of [`ModuleGraph`].
+///
+/// - `job_name` is just used for printing watcher status to terminal.
+///
+/// Note that the watcher will try to continue watching files using the previously resolved
+/// data if `module_resolver` fails at some point, which means the watcher won't work at all
+/// if `module_resolver` fails at the first attempt.
+///
+/// [`ModuleGraph`]: crate::module_graph::Graph
+/// [`ModuleSpecifier`]: deno_core::ModuleSpecifier
+pub async fn watch_func_with_module_resolution<F, G, T>(
+  module_resolver: F,
+  operation: G,
+  job_name: &str,
+) -> Result<(), AnyError>
+where
+  F: Fn() -> FileWatcherFuture<(Vec<PathBuf>, T)>,
+  G: Fn(T) -> FileWatcherFuture<()>,
+  T: Clone,
+{
+  let mut debounce = Debounce::new();
+  // Store previous data. If module resolution fails at some point, the watcher will try to
+  // continue watching files using these data.
+  let mut paths = None;
+  let mut module = None;
+
+  loop {
+    match module_resolver().await {
+      Ok((next_paths, next_module)) => {
+        paths = Some(next_paths);
+        module = Some(next_module);
+      }
+      Err(e) => {
+        // If at least one of `paths` and `module` is `None`, the watcher cannot decide which files
+        // should be watched. So return the error immediately without watching anything.
+        if paths.is_none() || module.is_none() {
+          return Err(e.into());
+        }
+      }
```

Comment:
To address #7848, I fixed `watch_func_with_module_resolution` implementation so that the watcher will try to continue watching files using _previously resolved_ data.

This works fine if the target files initially have correct syntax and then updated into invalid syntax.
This also means that the watcher won't work at all if the target files initially have invalid syntax. I think it cannot be helped because the watcher cannot decide which files should be watched in this case.

Reply:
If there's no problem with my approach, I'll add some integration tests to confirm this behavior. WDYT? @bartlomieju

Reply:
@magurotuna that sounds good! Test would be great